File: airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/FeatureFlagFetchActivity.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.temporal.scheduling.activities;

import io.temporal.activity.ActivityInterface;
import io.temporal.activity.ActivityMethod;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

/**
 * Activity that fetches feature flags.
 */
@ActivityInterface
public interface FeatureFlagFetchActivity {

  /**
   * Feature flag fetch input.
   */
  class FeatureFlagFetchInput {

    private UUID connectionId;

    public FeatureFlagFetchInput() {}

    public FeatureFlagFetchInput(UUID connectionId) {
      this.connectionId = connectionId;
    }

    public UUID getConnectionId() {
      return connectionId;
    }

    public void setConnectionId(UUID connectionId) {
      this.connectionId = connectionId;
    }

    @Override
    public boolean equals(Object o) {
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      FeatureFlagFetchInput that = (FeatureFlagFetchInput) o;
      return Objects.equals(connectionId, that.connectionId);
    }

    @Override
    public int hashCode() {
      return Objects.hashCode(connectionId);
    }

    @Override
    public String toString() {
      return "FeatureFlagFetchInput{connectionId=" + connectionId + '}';
    }

  }

  /**
   * Feature flag fetch output.
   */
  class FeatureFlagFetchOutput {

    private Map<String, Boolean> featureFlags;

    public FeatureFlagFetchOutput() {}

    public FeatureFlagFetchOutput(Map<String, Boolean> featureFlags) {
      this.featureFlags = featureFlags;
    }

    public Map<String, Boolean> getFeatureFlags() {
      return featureFlags;
    }

    public void setFeatureFlags(Map<String, Boolean> featureFlags) {
      this.featureFlags = featureFlags;
    }

    @Override
    public boolean equals(Object o) {
      if (o == null || getClass() != o.getClass()) {
        return false;
      }
      FeatureFlagFetchOutput that = (FeatureFlagFetchOutput) o;
      return Objects.equals(featureFlags, that.featureFlags);
    }

    @Override
    public int hashCode() {
      return Objects.hashCode(featureFlags);
    }

    @Override
    public String toString() {
      return "FeatureFlagFetchOutput{featureFlags=" + featureFlags + '}';
    }

  }

  /**
   * Return latest value for feature flags relevant to the ConnectionManagerWorkflow.
   */
  @ActivityMethod
  FeatureFlagFetchOutput getFeatureFlags(FeatureFlagFetchInput input);

}

File: airbyte-workers/src/main/kotlin/io/airbyte/workers/temporal/scheduling/activities/FeatureFlagFetchActivity.kt
/*
 * Copyright (c) 2020-2026 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.workers.temporal.scheduling.activities

import io.temporal.activity.ActivityInterface
import io.temporal.activity.ActivityMethod
import java.util.Objects
import java.util.UUID

/**
 * Activity that fetches feature flags.
 */
@ActivityInterface
interface FeatureFlagFetchActivity {
  /**
   * Feature flag fetch input.
   */
  class FeatureFlagFetchInput {
    @JvmField
    var connectionId: UUID? = null

    @JvmField
    var organizationId: UUID? = null

    constructor()

    constructor(connectionId: UUID?) {
      this.connectionId = connectionId
    }

    constructor(connectionId: UUID?, organizationId: UUID?) {
      this.connectionId = connectionId
      this.organizationId = organizationId
    }

    override fun equals(o: Any?): Boolean {
      if (o == null || javaClass != o.javaClass) {
        return false
      }
      val that = o as FeatureFlagFetchInput
      return connectionId == that.connectionId && organizationId == that.organizationId
    }

    override fun hashCode(): Int = Objects.hash(connectionId, organizationId)

    override fun toString(): String = "FeatureFlagFetchInput{connectionId=$connectionId, organizationId=$organizationId}"
  }

  /**
   * Feature flag fetch output.
   */
  class FeatureFlagFetchOutput {
    @JvmField
    var featureFlags: MutableMap<String?, Boolean?>? = null

    constructor()

    constructor(featureFlags: MutableMap<String?, Boolean?>?) {
      this.featureFlags = featureFlags
    }

    override fun equals(o: Any?): Boolean {
      if (o == null || javaClass != o.javaClass) {
        return false
      }
      val that = o as FeatureFlagFetchOutput
      return featureFlags == that.featureFlags
    }

    override fun hashCode(): Int = Objects.hashCode(featureFlags)

    override fun toString(): String = "FeatureFlagFetchOutput{featureFlags=" + featureFlags + '}'
  }

  /**
   * Return latest value for feature flags relevant to the ConnectionManagerWorkflow.
   */
  @ActivityMethod
  fun getFeatureFlags(input: FeatureFlagFetchInput): FeatureFlagFetchOutput
}
